src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -114,141 +162,125 @@ object ReadStructure {
+
+  def templateSegments:         Seq[ReadSegment] = segments(SegmentType.Template)
+  def sampleBarcodeSegments:    Seq[ReadSegment] = segments(SegmentType.SampleBarcode)
+  def molecularBarcodeSegments: Seq[ReadSegment] = segments(SegmentType.MolecularBarcode)
+  def skipSegments:             Seq[ReadSegment] = segments(SegmentType.Skip)
+}
+
+/** Sealed class hierarchy for the types of segments that can show up in a read structure. */
+sealed abstract class SegmentType(val code: Char)
+object SegmentType {
+  case object Template extends SegmentType('T')
+  case object SampleBarcode extends SegmentType('B')
+  case object MolecularBarcode extends SegmentType('M')
+  case object Skip extends SegmentType('S')
 
-  /** Creates a new read segment of the same type but with a new offset and length */
-  def apply(segment: ReadSegment, offset: Int, length: Int): ReadSegment = segment match {
-    case seg: Template         => Template(offset, length)
-    case seg: SampleBarcode    => SampleBarcode(offset, length)
-    case seg: MolecularBarcode => MolecularBarcode(offset, length)
-    case seg: Skip             => Skip(offset, length)
+  /** Returns the [[SegmentType]] for the given code/letter. */
+  def apply(code: Char): SegmentType = code match {
+    case 'T' => Template
+    case 'B' => SampleBarcode
+    case 'M' => MolecularBarcode
+    case 'S' => Skip
+    case _   => throw new IllegalArgumentException(s"Invalid read segment type: $code")
   }
 }
 
-/** Base for all types of segments of a read */
-sealed trait ReadSegment {
-  /** The number of bases in the read preceding this segment. */
-  def offset: Int
-  /** The length of this segment. */
-  def length: Int
-  /** The single-character symbol representing this segment. */
-  def symbol: Char
-
-  /** Gets first and last bases (exclusive, 0-based) associated with this read segment.  If strict is false then only
-    * bound the last base to `basesLen`, otherwise if the end is too big, throw an exception.
-    **/
-  private[util] def range(basesLen: Int, strict: Boolean): (Int, Int) = {
-    val start = this.offset
-    val end = length + start
-    if (basesLen < end) {
-      if (strict) throw new IllegalStateException("Not enough bases for read segment: " + this)
-      else (start, basesLen)
-    }
-    else (start, end)
+object ReadSegment {
+  val Types = Seq('T', 'B', 'M', 'S')
+
+  /** Creates a new ReadSegment with undefined length (i.e. length=0 or more). */
+  def apply(offset: Int, c: Char): ReadSegment = new ReadSegment(offset, None, kind=SegmentType(c))
+
+  /** Constructs a ReadSegment with a definite length using Char c to find the segment type. */
+  def apply(offset: Int, length: Int, c: Char): ReadSegment = new ReadSegment(offset=offset, length=Some(length), kind=SegmentType(c))
+}
+
+/** Encapsulates all the information about a segment within a read structure. */
+case class ReadSegment (offset: Int, length: Option[Int], kind: SegmentType) {
```

Comment:
Good call.  I've changed the `hasDefiniteLength` to `hasFixedLength`.  I originally thought the former was a good idea since it mimic'd what scala collections use, but that's really for distinguishing finite vs. infinite not exact vs. variable, so I think `fixed` is better.  I've added `fixedLength` and used it where appropriate.